src/core/mount.c

```diff
@@ -299,26 +299,24 @@ static int mount_add_mount_links(Mount *m) {
                 if (!parent)
                         return -ENOMEM;
 
-                r = unit_require_mounts_for(UNIT(m), parent);
+                r = unit_require_mounts_for(UNIT(m), parent, UNIT_DEPENDENCY_IMPLICIT);
                 if (r < 0)
                         return r;
         }
 
-        /* Adds in links to other mount points that might be needed
-         * for the source path (if this is a bind mount or a loop mount) to be
-         * available. */
+        /* Adds in dependencies to other mount points that might be needed for the source path (if this is a bind mount
+         * or a loop mount) to be available. */
         pm = get_mount_parameters_fragment(m);
         if (pm && pm->what &&
             path_is_absolute(pm->what) &&
             (mount_is_bind(pm) || mount_is_loop(pm) || !mount_is_network(pm))) {
 
-                r = unit_require_mounts_for(UNIT(m), pm->what);
+                r = unit_require_mounts_for(UNIT(m), pm->what, UNIT_DEPENDENCY_FILE);
```

Comment:
U_D_FILE matches the comment for U_D_FILE in unit.h. Nevertheless, U_D_IMPLICIT would seem more natural. Why is it _FILE here and _IMPLICIT in line 302?

Reply:
it's all depending on where the dependency was configured: deps that are created exclusively based on the name of the unit, are of the "implicit" kind, and that's the case for the dep in 302: we check m->where, which in turn is derived directly from the unit name. That means even if the unit had no unit file at all, the dep on its parent would be created. Or to say this differently: regardless what the unit file defines, the parent deps is always created, hence it is implicit.

This is different for the dep in 314: it depends on stuff that is configured in the unit file, as we use get_mount_parameters_fragment() and then use pm->what from that. If the unit file would not exist at all (maybe because the mount unit appeared just from /proc/self/mountinfo), get_mount_parameters_fragment() would return NULL, and no dep of this kind would be created. Moreover, depending on the contents of the unit file pm->what differ greatly, hence this dep is created as UNIT_DEPENDENCY_FILE.

Or to see this from a different standpoint: let's say we implement "pinpointed" reloads, i.e. we want to flush out every single bit of config of the unit that was generated from the old unit file, and then load the new unit file with the new config. When flushing out, we'd like to remove the deps created due to pm->what in preparation that it might change with the new unit file. OTOH the dep for m->what we don't have to flush out, since the it only depends on the name of the unit, and that's not gonna change during the reload. Hence the latter is marked as _IMPLICIT, the other AS _FILE

Reply:
OK, makes sense.